src/apps/rigoblock/common/rigoblock.pool.contract-position-fetcher.ts

```diff
@@ -0,0 +1,147 @@
+import { UniswapV3ContractFactory } from '~apps/uniswap-v3/contracts';
+import { UniswapV3LiquidityContractPositionBuilder } from '~apps/uniswap-v3/common/uniswap-v3.liquidity.contract-position-builder';
+import { UniswapV3LiquidityPositionDataProps } from '~apps/uniswap-v3/common/uniswap-v3.liquidity.contract-position-fetcher';
+import { ContractPositionBalance } from '~position/position-balance.interface';
+import { MetaType } from '~position/position.interface';
+import { GetDataPropsParams } from '~position/template/contract-position.template.types';
+import { ContractPositionTemplatePositionFetcher } from '~position/template/contract-position.template.position-fetcher';
+import { RigoblockContractFactory, SmartPool } from '../contracts';
+
+export type UnderlyingLiquidityPositionTokens = {
+  address: string;
+  balance: number;
+  balanceRaw: string;
+  balanceUSD: number;
+}
+
+export type RigoblockLiquidityDataProps = {
+  liquidity: number;
+  liquidityPositions: UnderlyingLiquidityPositionTokens[];
+};
+
+export type RigoblockPoolAppTokenDefinition = {
+  address: string;
+  label: string;
+};
+
+export abstract class RigoblockPoolContractPositionFetcher extends ContractPositionTemplatePositionFetcher<
+  SmartPool,
+  RigoblockLiquidityDataProps,
+  RigoblockPoolAppTokenDefinition
+> {
+  abstract positionManagerAddress: string;
+  abstract groupLabel: string;
+
+  constructor(
+    @Inject(APP_TOOLKIT) protected readonly appToolkit: IAppToolkit,
+    @Inject(RigoblockContractFactory) protected readonly rigoblockContractFactory: RigoblockContractFactory,
+    @Inject(UniswapV3ContractFactory) protected readonly contractFactory: UniswapV3ContractFactory,
+    @Inject(UniswapV3LiquidityContractPositionBuilder)
+    protected readonly uniswapV3LiquidityContractPositionBuilder: UniswapV3LiquidityContractPositionBuilder,
+  ) {
+    super(appToolkit);
+  }
+
+  getContract(address: string): SmartPool {
+    return this.rigoblockContractFactory.smartPool({ address, network: this.network });
+  }
+
+  async getDefinitions(): Promise<RigoblockPoolAppTokenDefinition[]> {
+    const appTokens = await this.appToolkit.getAppTokenPositions({
+      appId: this.appId,
+      network: this.network,
+      groupIds: ['pool'],
+    });
+
+    return appTokens.map(pool => {
+      return {
+        address: pool.address,
+        label: pool.displayProps.label,
+      };
+    });
+  }
+
+  // we define the liquidity position of a rigoblock pool
+  async getTokenDefinitions({ definition }) {
+    const underlyingTokenBalances = await this.getUnderlyingBalances(definition.address);
+    return underlyingTokenBalances.map(balance => {
+      const tokens = balance.tokens;
+      return tokens.map(token => {
+        return [
+          {
+            metaType: MetaType.SUPPLIED,
+            address: token.address,
+            network: this.network,
+          },
+        ];
+      });
+    }).flat(2);
+  }
```

Comment:
I suggest you implement the logic fetching the underlying tokens in the `getDefinitions` since it's freee. 

You can get the underlying token addresses by  `appTokens.map(x => x.tokens.map(y => y.address))`

Reply:
This has been done. I do not later retrieve from appToken.tokens as would require an extra RPC call. I am storing the tokens and the balances by making the onchain calls in the definitions, then saving some data that is needed in the application in dataProps and reading them from there.

Reply:
I'm not sure I understand why you're fetching everything in the definition, propagating it all the way down to method which retrieve balances ?

Reply:
The getBalances method only accepts address: string as input (please correct me if I am wrong as I would be more than happy to pass contractPosition as input in getBalances using a custom contract position fetcher to save the rpc calls), so regardless of it being a contract position template or overwriting it with a custom contract position template the onchain call is required at that level. However, in order to define the liquidity positions owned by each rigoblock pool, I have to make the same onchain call, as with the same call both underlying liquidity position tokens and balances are returned.

Reply:
When doing a bunch of read with Multicall, there's almost no need to optimize the number of RPC calls. In getTokenDefinitions, you don't need to have the actual balances of a pool, all you need are token addresses. `GetBalance` has to return the balance for a specific wallet and not for a pool (I might be confused as I'm not familiar with rigoblock)